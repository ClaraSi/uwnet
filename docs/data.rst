.. currentmodule:: uwnet

Data Loading
============

This package uses pytorch's classes to load the data into mini-batches which are suitable for training neural networks.


API
---

.. automodule:: uwnet.data
   :members:
   :undoc-members:
