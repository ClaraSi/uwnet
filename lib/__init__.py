from .interp import pressure_interp_ds
