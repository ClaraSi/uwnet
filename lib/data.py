from glob import glob

import numpy as np

import xarray as xr
from xnoah import centderiv

from .advection import material_derivative
from .thermo import (d_liquid_water_temperature, layer_mass, lhf_to_tendency,
                     liquid_water_temperature, shf_to_tendency, total_water)
from .util import compute_weighted_scale


def prepare_array(x):
    output_dims = [dim for dim in ['time', 'y', 'x', 'z'] if dim in x.dims]
    return x.transpose(*output_dims).values


def prepare_data(inputs: xr.Dataset, forcings: xr.Dataset):

    w = inputs.w

    fields = ['sl', 'qt']

    weights = {key: w.values for key in fields}

    # compute scales
    sample_dims = set(['x', 'y', 'time']) & set(inputs.dims)
    scales = compute_weighted_scale(
        w, sample_dims=sample_dims, ds=inputs[fields])
    scales = {key: float(scales[key]) for key in fields}

    X = {key: prepare_array(inputs[key]) for key in inputs.data_vars}
    G = {key: prepare_array(forcings[key]) for key in forcings.data_vars}

    # return stacked data

    return {
        'X': X,
        'G': G,
        'scales': scales,
        'w': weights,
        'p': inputs.p.values,
        'z': inputs.z.values
    }


def defaultsel(x):
    return x.isel(y=slice(24, 40))


def inputs_and_forcings(file_3d, file_2d, stat_file, sel=defaultsel):

    data_3d = xr.open_dataset(file_3d)
    data_2d = xr.open_dataset(file_2d)
    stat = xr.open_dataset(stat_file)

    data = xr.merge((data_3d, data_2d), join='inner').sortby('time')

    # limit to tropics
    data = sel(data)

    p = stat.p
    rho = stat.RHO[0].drop('time')
    w = layer_mass(rho)

    # compute the surface flux tendencies

    inputs = xr.Dataset({
        'sl':
        liquid_water_temperature(data.TABS, data.QN, data.QP),
        'qt':
        total_water(data.QV, data.QN)
    })

    forcings = inputs.apply(
        lambda f: -material_derivative(data.U, data.V, data.W, f) * 86400)

    forcings['SHF'] = data.SHF
    forcings['LHF'] = data.LHF
    forcings['QRAD'] = data.QRAD
    forcings['Prec'] = data.Prec
    forcings['W'] = data.W
    forcings['SOLIN'] = data.SOLIN

    inputs['p'] = p
    inputs['w'] = w

    return inputs, forcings
