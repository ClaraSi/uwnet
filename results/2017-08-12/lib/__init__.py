import numpy as np
from scipy.linalg import pinv
import gnl.spline as gs
from gnl.xarray import xr
from glob import glob


def smooth(A, dim, n, per=True):
    x = np.asarray(A[dim])

    L = x[-1]*2 - x[-2] if per else x[-1]
    knots = np.linspace(x[0], L, n)

    if per:
        Bx = gs.psplines(x, knots)
    else:
        Bx = gs.splines(x, knots)

    new_dim = 'poksd'
    S = pinv(Bx) @ Bx
    S = xr.DataArray(S, dims=[dim, new_dim], coords={dim: A[dim], new_dim: np.asarray(A[dim])})
    return A.dot(S).rename({new_dim:dim})


class SamRun(object):
    def __init__(self, rev, root="/scratch/ndb245/Data/SAM6.10.9/JOBS"):
        g = glob("{}/*-{}*".format(root, rev))

        if len(g) == 0:
            raise ValueError("Specified ID not found")
        elif len(g) > 1:
            raise ValueError("Specified ID is not unique")
        else:
            self.rev = g[0]

    @property
    def data2d(self):
        return self.ncglob("OUT_2D")

    @property
    def data3d(self):
        return self.ncglob("OUT_3D")

    @property
    def moments(self):
        return self.ncglob("OUT_MOMENTS")
    @property
    def stats(self):
        return self.ncglob("OUT_STAT")



    def ncglob(self,  path):
        return sorted(glob("{}/{}/*.nc".format(self.rev, path)))
