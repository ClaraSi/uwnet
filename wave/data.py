import intake

from .thermo import compute_apparent_source

TRAINING_DATA_PATH = "gs://vcm-ml-data/project_data/uwnet/2020-02-01-noBlur.zarr/"


def open_data(tag):
    if tag == 'training':
        return intake.open_zarr(TRAINING_DATA_PATH).to_dask()


def assign_apparent_sources(ds):
    return ds.assign(
        Q1=compute_apparent_source(ds.SLI, 86400 * ds.FSLI),
        Q2=compute_apparent_source(ds.QT, 86400 * ds.FQT))
