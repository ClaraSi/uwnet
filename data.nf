#!/usr/bin/env nextflow
// -*- mode: groovy; -*-
/*
========================================================================================
                         sam-uwnet
 ========================================================================================
 sam-uwnet Analysis Pipeline. Started 2018-08-02.
 #### Homepage / Documentation
 https://github.com/sam-uwnet
 #### Authors
 Noah D. Brenowitz nbren12 <nbren12@uw.edu> - http://www.noahbrenowitz.com
----------------------------------------------------------------------------------------
*/

params.numTimePoints = 640
params.NGAquaDir = "/Users/noah/Data/2018-05-30-NG_5120x2560x34_4km_10s_QOBS_EQX"

/*
 Process the NGAqua data
 */

time_step_ch = Channel.from(0..params.numTimePoints - 1)
stat_file_ch = Channel.fromPath(params.NGAquaDir + "/stat.nc")
sfc_file_ch = Channel.fromPath(params.NGAquaDir + "/coarse/2d/all.nc")


process processOneTimeStep {
    input:
        val t from time_step_ch

    output:
        file 'rec.*.nc' into single_time_steps_ch

    """
    process_ngaqua.py -n $params.NGAquaDir $t
    f=\$(ls *.nc)
    ncks -A --mk_rec_dmn time \$f rec.\$f
    rm -f \$f
    """
}


process concatAllFiles {
    input:
        file x from single_time_steps_ch.collect()
    output:
        file 'training_data.nc' into concated_file_ch
    """
    echo $x | tr ' ' '\\n' | sort -n | ncrcat -o training_data.nc
    ncatted -a units,FQT,c,c,'g/kg/s' \
            -a units,FSLI,c,c,'K/s' \
            -a units,FU,c,c,'m/s^2' \
            -a units,FV,c,c,'m/s^2' \
            -a units,x,c,c,'m' \
            -a units,y,c,c,'m' \
             training_data.nc
    """
}


process addOtherNeededVariables {
    input:
        file x from concated_file_ch
        file stat from stat_file_ch
        file d2d from sfc_file_ch
    output:
        set file('aux.nc'), file(x) into combine_vars_ch

    """
    #!/usr/bin/env python
    import xarray as xr
    from uwnet.thermo import layer_mass

    # compute layer_mass
    stat = xr.open_dataset("$stat")
    rho = stat.RHO.isel(time=0).drop('time')
    w = layer_mass(rho)

    # get 2D variables
    d2 = xr.open_dataset("$d2d")
    ds = xr.open_dataset("$x")
    d2 = d2.sel(time=ds.time)

    out = xr.Dataset({
       'RADTOA': d2.LWNT - d2.SWNT,
       'RADSFC': d2.LWNS - d2.SWNS,
       'layer_mass': w
    })


    for key in 'Prec SHF LHF SOLIN SST'.split():
        out[key] = d2[key]

    # append these variables
    out.to_netcdf("aux.nc")

    """
}

process combineVariables {
    publishDir 'data'
    input:
        set file(x), file(y) from combine_vars_ch
    output:
        file y into training_data_ch
    """
    ncks -A $x $y
    """
}

process subsetData {
    publishDir 'data'
    input:
    file x from training_data_ch

    output:
    file 'subset.nc'
    """
    ncks -d time,0,200 -d x,0,16 $x subset.nc
    """
}
