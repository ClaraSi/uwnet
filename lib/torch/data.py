"""This module contains routines for handling the intermediate representation
of the data. The routines in this module all have a dict `data` as the first
argument. This dict should have the same structure as produced by `prepare_data`.

Training.py uses the following interface:

to_dataset
to_constants ...done
to_scaler ...done
to_dynamics_loss

the useful parts of ..data are

compute_weighted_scale : computes the scales
prepare_array : transposes the data correctly
the material derivative part of inputs_and_forcings



See Also
--------
lib.data.prepare_data

"""
import logging

import attr
import numpy as np
import torch
from toolz import valmap, curry
from torch.utils.data import DataLoader
from torch.utils.data.dataloader import default_collate
from torch.utils.data.sampler import SubsetRandomSampler

import xarray as xr

from ..thermo import layer_mass, liquid_water_temperature
from .datasets import DictDataset, WindowedData
from .loss import dynamic_loss
from .normalization import scaler

logger = logging.getLogger(__name__)



def _compute_weighted_scale(x, layer_mass):
    sample_dims = tuple(range(x.ndim - 1))
    sig2 = x.var(axis=sample_dims)

    assert sig2.shape == layer_mass.shape

    avg_sig2 = (sig2 * layer_mass / layer_mass.sum()).sum()
    return np.sqrt(avg_sig2)


def _loss_weight(x, layer_mass):
    """Compute MSE loss weight for numpy array

    Parameters
    ----------
    x : (..., z)
    layer_mass : (z, )
    """
    scale = _compute_weighted_scale(x, layer_mass)
    return layer_mass / scale**2


def prepare_array(x):
    output_dims = [dim for dim in ['time', 'y', 'x', 'z'] if dim in x.dims]
    return x.transpose(*output_dims).values.astype(np.float32)


@curry
def collate_fn(constants, seq):
    collated = default_collate(seq)
    d = _transpose_vars(collated)
    d['constant'] = constants
    return d


def _transpose_vars(data):
    if torch.is_tensor(data):
        return data.transpose(0, 1)
    elif isinstance(data, dict):
        return {
            key: _transpose_vars(val)
            for key, val in data.items()
        }


@attr.s
class TrainingData(object):
    qt = attr.ib()
    sl = attr.ib()
    FQT = attr.ib()
    FSL = attr.ib()
    SHF = attr.ib()
    LHF = attr.ib()
    QRAD = attr.ib()
    SOLIN = attr.ib()
    layer_mass = attr.ib()
    z = attr.ib()
    p = attr.ib()

    def constants(self):
        return {
            'w': self.layer_mass,
            'z': self.z
        }

    def prognostic(self):
        return {
            'qt': self.qt,
            'sl': self.sl,
        }

    def forcing(self):
        # create forcing dictionary
        auxiliary_names = ('QRAD', 'LHF', 'SHF', 'SOLIN')
        forcing_variables = {
            'sl': self.FSL,
            'qt': self.FQT,
        }
        for name in auxiliary_names:
            forcing_variables[name] = getattr(self, name)

        return forcing_variables

    def scaler(self):
        # compute mean and stddev
        # this is an error, std does not work like this
        means = {}
        scales = {}
        input_vars = {'qt': self.qt, 'sl': self.sl}

        for key, X in input_vars.items():
            m = X.shape[-1]
            mu = X.reshape((-1, m)).mean(axis=0)
            sig = X.reshape((-1, m)).std(axis=0)

            # convert to torch
            mu, sig = [torch.from_numpy(np.squeeze(x)) for x in [mu, sig]]

            # take average in vertical direction
            sig = torch.mean(sig)

            means[key] = mu
            scales[key] = sig

        return scaler(scales, means)

    def dynamic_loss(self, **kwargs):
        weights = {
            'qt': _loss_weight(self.qt, self.layer_mass),
            'sl': _loss_weight(self.sl, self.layer_mass),
        }
        weights = valmap(torch.from_numpy, weights)
        return dynamic_loss(weights=weights, **kwargs)

    def get_num_features(self):
        return len(self.p) * 2

    def torch_dataset(self, window_size):

        # create prognostic dict
        prognostic_variables = {'sl': self.sl, 'qt': self.qt}

        # create forcing dictionary
        auxiliary_names = ('QRAD', 'LHF', 'SHF', 'SOLIN')
        forcing_variables = {
            'sl': self.FSL,
            'qt': self.FQT,
        }
        for name in auxiliary_names:
            forcing_variables[name] = getattr(self, name)

        # turn these into windowed datasets
        X = DictDataset({
            key: WindowedData(val, window_size)
            for key, val in prognostic_variables.items()
        })
        G = DictDataset({
            key: WindowedData(val, window_size)
            for key, val in forcing_variables.items()
        })

        return DictDataset({'prognostic': X, 'forcing': G})

    def get_loader(self, window_size, num_samples=None, batch_size=None,
                   **kwargs):
        """Return the whole dataset as a batch for input to ForcedStepper

        Yields
        -------
        batch : dict
            contains keys 'constant', 'prognostic', 'forcing'.

        """
        dataset = self.torch_dataset(window_size)


        # Create training data loaders
        if num_samples:
            logger.info(f"Using boostrap sample of {num_samples}.")
            inds = np.random.choice(len(dataset), num_samples,
                                    replace=False)
            kwargs['sampler'] = SubsetRandomSampler(inds)
        else:
            logger.info(f"Using full training dataset")
            num_samples = len(dataset)

        if batch_size:
            kwargs['batch_size'] = batch_size
        else:
            kwargs['batch_size'] = num_samples

        constants = valmap(torch.from_numpy, self.constants())
        return DataLoader(dataset, collate_fn=collate_fn(constants),
                          **kwargs)


    def from_files(paths, post=None):
        """Create TrainingData from filenames

        Examples
        --------
        >>> TrainingData.from_files(default='all.nc', p='p.nc')
        """
        init_kwargs = {}
        for key in paths:
            logger.debug(f"Loading {paths[key]}")
            init_kwargs[key] = xr.open_dataset(
                paths[key], chunks={
                    'time': 10
                })[key]

        # compute layer mass from stat file
        if 'RHO' in paths:
            rho = init_kwargs.pop('RHO')[0]
            rhodz = layer_mass(rho)

            init_kwargs['layer_mass'] = rhodz
            init_kwargs['z'] = rhodz.z

        # compute prognostics variables from TABS
        if 'TABS' in paths:
            TABS = init_kwargs.pop('TABS')
            QV = init_kwargs.pop('QV')
            QN = init_kwargs.pop('QN', 0.0)
            QP = init_kwargs.pop('QP', 0.0)

            sl = liquid_water_temperature(TABS, QN, QP)
            qt = QV + QN

            init_kwargs['sl'] = sl
            init_kwargs['qt'] = qt

        if post:
            init_kwargs = valmap(post, init_kwargs)

        # process all files into numpy arrays
        init_kwargs = {key: prepare_array(x) for key, x in init_kwargs.items()}

        return TrainingData(**init_kwargs)

    @classmethod
    def from_var_files(cls, files, **kwargs):
        paths = {}
        for path, variables in files:
            paths.update({key: path for key in variables})

        return cls.from_files(paths, **kwargs)

    def save(self, path):
        torch.save(attr.asdict(self), path)

    @classmethod
    def load(cls, path):
        return cls(**torch.load(path))

    @property
    def nbytes(self):
        total = 0
        for val in attr.astuple(self):
            total += val.nbytes
        return total

    @property
    def shape(self):
        return self.FQT.shape
