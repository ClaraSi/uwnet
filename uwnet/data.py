import click
import numpy as np
import torch
from dask.diagnostics import ProgressBar
from toolz import valmap
from torch.utils.data import Dataset

import xarray as xr
from uwnet import thermo


def _stack_or_rename(x, **kwargs):
    for key, val in kwargs.items():
        if isinstance(val, str):
            x = x.rename({val: key})
        else:
            x = x.stack(**{key: val})
    return x


def _ds_slice_to_numpy_dict(ds):
    out = {}
    for key in ds.data_vars:
        out[key] = _to_numpy(ds[key])
    return out


def _to_numpy(x: xr.DataArray):
    dim_order = ['xbatch', 'xtime', 'xfeat']
    dims = [dim for dim in dim_order if dim in x.dims]
    return x.transpose(*dims).values


def _ds_slice_to_torch(ds):
    return valmap(lambda x: torch.from_numpy(x).detach(),
                  _ds_slice_to_numpy_dict(ds))


class XRTimeSeries(Dataset):
    """A pytorch Dataset class for time series data in xarray format

    Parameters
    ----------
    data : xr.Dataset
        input data
    dims : seq
        list of dimensions used to reshape the data. Format::

            (time_dims, batch_dims, feature_dims)

    Attributes
    ----------
    std
    mean
    scale

    Examples
    --------
    >>> ds = xr.open_dataset("in.nc")
    >>> XRTimeSeries(ds, [['time'], ['x', 'y'], ['z']])

    """

    def __init__(self, data, dims):
        """Initialize XRTimeSeries.

        """
        self.data = data
        self.dims = dims
        self._ds = _stack_or_rename(
            self.data,
            xtime=self.dims[0],
            xbatch=self.dims[1],
            xfeat=self.dims[2])

    def __len__(self):
        res = 1
        for dim in self.dims[1]:
            res *= len(self.data[dim])
        return res

    def __getitem__(self, i):

        # convert i to an array
        # this code should handle i = slice, list, etc
        i = np.arange(len(self))[i]
        scalar_idx = i.ndim == 0
        if scalar_idx:
            i = [i]

        # get coordinates using np.unravel_index
        # this code should probably be refactored
        batch_dims = self.dims[1]
        batch_shape = [len(self.data[dim]) for dim in batch_dims]

        idxs = np.unravel_index(i, batch_shape)
        coords = {}
        for key, idx in zip(batch_dims, idxs):
            coords[key] = xr.DataArray(idx, dims='xbatch')

        # select, load, and stack the batch
        batch_ds = self.data.isel(**coords).load()
        ds_r = _stack_or_rename(
            batch_ds, xtime=self.dims[0], xfeat=self.dims[2])

        # prepare for output
        out = {}
        for key in ds_r.data_vars:
            if key in self.constants():
                continue

            arr = _to_numpy(ds_r[key])
            arr = torch.from_numpy(arr)
            arr.requires_grad = False
            if scalar_idx:
                arr = arr[0]

            out[key] = arr.float()

        return out

    @property
    def time_dim(self):
        return self.dims[0][0]

    def constants(self):
        for key in self.data:
            if self.time_dim not in self.data[key].dims:
                yield key

    def torch_constants(self):
        return {
            key: torch.tensor(self.data[key].values, requires_grad=False)
            .float()
            for key in self.constants()
        }

    @property
    def mean(self):
        """Mean of the contained variables"""
        ds = self._ds.mean(['xbatch', 'xtime'])
        return _ds_slice_to_torch(ds)

    @property
    def std(self):
        """Standard deviation of the contained variables"""
        ds = self._ds.std(['xbatch', 'xtime'])
        return _ds_slice_to_torch(ds)

    @property
    def scale(self):
        std = self.std
        return valmap(lambda x: x.max(), std)

    def timestep(self):
        time_dim = self.dims[0][0]
        time = self.data[time_dim]
        dt = np.diff(time)

        all_equal = dt.std() / dt.mean() < 1e-6
        if not all_equal:
            raise ValueError("Data must be uniformly sampled in time")

        return dt[0]


def compute_radiation(data):
    lw = data.pop('LWNS')
    sw = data.pop('SWNS')
    data['RADSFC'] = lw - sw

    lw = data.pop('LWNT')
    sw = data.pop('SWNT')
    data['RADTOA'] = lw - sw

    return data


def compute_forcings(data):
    pass


def load_data(paths):

    data = {}
    for info in paths:
        for field in info['fields']:
            try:
                ds = xr.open_dataset(info['path'], chunks={'time': 40})
            except ValueError:
                ds = xr.open_dataset(info['path'])

            data[field] = ds[field]

    # compute layer mass from stat file
    # remove presssur
    rho = data.pop('RHO')[0]
    rhodz = thermo.layer_mass(rho)

    data['layer_mass'] = rhodz

    # compute thermodynamic variables
    TABS = data.pop('TABS')
    QV = data.pop('QV')
    QN = data.pop('QN', 0.0)
    QP = data.pop('QP', 0.0)

    sl = thermo.liquid_water_temperature(TABS, QN, QP)
    qt = QV + QN

    data['sl'] = sl.assign_attrs(units='K')
    data['qt'] = qt.assign_attrs(units='g/kg')

    data = compute_radiation(data)
    compute_forcings(data)

    # rename staggered dimensions
    data['U'] = (data['U']
                 .rename({'xs': 'x', 'yc': 'y'}))

    data['V'] = (data['V']
                 .rename({'xc': 'x', 'ys': 'y'}))

    objects = [
        val.to_dataset(name=key).assign(x=sl.x, y=sl.y)
        for key, val in data.items()
    ]

    ds = xr.merge(objects, join='inner').sortby('time')
    return ds




def get_dataset(paths, post=None, **kwargs):
    # paths = yaml.load(open("config.yaml"))['paths']
    ds = load_data(paths)
    if post is not None:
        ds = post(ds)
    return XRTimeSeries(ds, [['time'], ['x', 'y'], ['z']])


@click.command()
@click.argument("config")
@click.argument("out")
def main(config, out):
    import yaml
    paths = yaml.load(open(config))['paths']
    with ProgressBar():
        load_data(paths)\
            .load()\
            .to_zarr(out)


if __name__ == '__main__':
    main()
