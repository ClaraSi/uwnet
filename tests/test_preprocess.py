import numpy as np
import xarray as xr
from lib.data import prepare_data


def test_prepare_data(regtest):
    # mock the xarray inputs
    time, z, y, x = np.mgrid[:3, :3, :3, :3]

    dims = ['time', 'z', 'y', 'x']
    coords = {key: np.arange(3) for key in dims}


    inputs = xr.Dataset(
        {
            'qt': (dims, x + y),
            'sl': (dims, x + z)
        }, coords=coords)

    forcings = xr.Dataset(
        {
            'qt': (dims, x * y),
            'sl': (dims, x * z)
        }, coords=coords)

    w = - inputs.z

    inputs = inputs.assign(w=w, p=w)

    data = prepare_data(inputs, forcings)

    print(data, file=regtest)
