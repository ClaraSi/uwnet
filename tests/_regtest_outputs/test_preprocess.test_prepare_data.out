{'X': {'qt': array([[[[0, 0, 0],
         [1, 1, 1],
         [2, 2, 2]],

        [[1, 1, 1],
         [2, 2, 2],
         [3, 3, 3]],

        [[2, 2, 2],
         [3, 3, 3],
         [4, 4, 4]]],


       [[[0, 0, 0],
         [1, 1, 1],
         [2, 2, 2]],

        [[1, 1, 1],
         [2, 2, 2],
         [3, 3, 3]],

        [[2, 2, 2],
         [3, 3, 3],
         [4, 4, 4]]],


       [[[0, 0, 0],
         [1, 1, 1],
         [2, 2, 2]],

        [[1, 1, 1],
         [2, 2, 2],
         [3, 3, 3]],

        [[2, 2, 2],
         [3, 3, 3],
         [4, 4, 4]]]]), 'sl': array([[[[0, 1, 2],
         [1, 2, 3],
         [2, 3, 4]],

        [[0, 1, 2],
         [1, 2, 3],
         [2, 3, 4]],

        [[0, 1, 2],
         [1, 2, 3],
         [2, 3, 4]]],


       [[[0, 1, 2],
         [1, 2, 3],
         [2, 3, 4]],

        [[0, 1, 2],
         [1, 2, 3],
         [2, 3, 4]],

        [[0, 1, 2],
         [1, 2, 3],
         [2, 3, 4]]],


       [[[0, 1, 2],
         [1, 2, 3],
         [2, 3, 4]],

        [[0, 1, 2],
         [1, 2, 3],
         [2, 3, 4]],

        [[0, 1, 2],
         [1, 2, 3],
         [2, 3, 4]]]]), 'p': array([ 0, -1, -2]), 'w': array([ 0, -1, -2])}, 'G': {'qt': array([[[[0, 0, 0],
         [0, 0, 0],
         [0, 0, 0]],

        [[0, 0, 0],
         [1, 1, 1],
         [2, 2, 2]],

        [[0, 0, 0],
         [2, 2, 2],
         [4, 4, 4]]],


       [[[0, 0, 0],
         [0, 0, 0],
         [0, 0, 0]],

        [[0, 0, 0],
         [1, 1, 1],
         [2, 2, 2]],

        [[0, 0, 0],
         [2, 2, 2],
         [4, 4, 4]]],


       [[[0, 0, 0],
         [0, 0, 0],
         [0, 0, 0]],

        [[0, 0, 0],
         [1, 1, 1],
         [2, 2, 2]],

        [[0, 0, 0],
         [2, 2, 2],
         [4, 4, 4]]]]), 'sl': array([[[[0, 0, 0],
         [0, 1, 2],
         [0, 2, 4]],

        [[0, 0, 0],
         [0, 1, 2],
         [0, 2, 4]],

        [[0, 0, 0],
         [0, 1, 2],
         [0, 2, 4]]],


       [[[0, 0, 0],
         [0, 1, 2],
         [0, 2, 4]],

        [[0, 0, 0],
         [0, 1, 2],
         [0, 2, 4]],

        [[0, 0, 0],
         [0, 1, 2],
         [0, 2, 4]]],


       [[[0, 0, 0],
         [0, 1, 2],
         [0, 2, 4]],

        [[0, 0, 0],
         [0, 1, 2],
         [0, 2, 4]],

        [[0, 0, 0],
         [0, 1, 2],
         [0, 2, 4]]]])}, 'scales': {'sl': 0.816496580927726, 'qt': 1.1547005383792515}, 'w': {'sl': array([ 0, -1, -2]), 'qt': array([ 0, -1, -2])}, 'p': array([ 0, -1, -2]), 'z': array([0, 1, 2])}
